core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1331 @@
+    /**
+     * 10^x. All unsigned values.
+     */
+    private static final int[] POWERS_OF_TEN_INT = new int[MAX_POWER_OF_TEN_INT + 1];
+
+    static {
+        for (int i = 0; i < POWERS_OF_FIVE.length; ++i) {
+            POWERS_OF_FIVE[i] = Int128.valueOf(BigInteger.valueOf(5).pow(i));
+        }
+        for (int i = 0; i < POWERS_OF_TEN.length; ++i) {
+            POWERS_OF_TEN[i] = Int128.valueOf(BigInteger.TEN.pow(i));
+        }
+
+        POWERS_OF_FIVES_INT[0] = 1;
+        for (int i = 1; i < POWERS_OF_FIVES_INT.length; ++i) {
+            POWERS_OF_FIVES_INT[i] = POWERS_OF_FIVES_INT[i - 1] * 5;
+        }
+
+        POWERS_OF_FIVE_LONG[0] = 1;
+        for (int i = 1; i < POWERS_OF_FIVE_LONG.length; ++i) {
+            POWERS_OF_FIVE_LONG[i] = POWERS_OF_FIVE_LONG[i - 1] * 5;
+        }
+
+        POWERS_OF_TEN_INT[0] = 1;
+        for (int i = 1; i < POWERS_OF_TEN_INT.length; ++i) {
+            POWERS_OF_TEN_INT[i] = POWERS_OF_TEN_INT[i - 1] * 10;
+        }
+
+        if (!ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+            throw new IllegalStateException("UnsignedDecimal128Arithmetic is supported on little-endian machines only");
+        }
+    }
+
+    public static void rescale(long high, long low, int factor, long[] result, int offset)
+    {
+        if (factor == 0) {
+            result[0] = high;
+            result[1] = low;
+        }
+        else if (factor > 0) {
+            shiftLeftBy10(high, low, factor, result, offset);
+        }
+        else {
+            scaleDownRoundUp(high, low, -factor, result, offset);
+        }
+    }
+
+    public static Int128 rescale(Int128 decimal, int rescaleFactor)
+    {
+        long[] result = new long[2];
+        rescale(decimal.getHigh(), decimal.getLow(), rescaleFactor, result, 0);
+        return Int128.valueOf(result);
+    }
+
+    public static void rescaleTruncate(long high, long low, int rescaleFactor, long[] result, int offset)
+    {
+        if (rescaleFactor == 0) {
+            result[0] = high;
+            result[1] = low;
+        }
+        else if (rescaleFactor > 0) {
+            shiftLeftBy10(high, low, rescaleFactor, result, offset);
+        }
+        else {
+            scaleDownTruncate(high, low, -rescaleFactor, result, offset);
+        }
+    }
+
+    public static Int128 rescaleTruncate(Int128 decimal, int rescaleFactor)
+    {
+        long[] result = new long[2];
+        rescaleTruncate(decimal.getHigh(), decimal.getLow(), rescaleFactor, result, 0);
+        return Int128.valueOf(result);
+    }
+
+    // Multiplies by 10^rescaleFactor. Only positive rescaleFactor values are allowed
+    private static void shiftLeftBy10(long high, long low, int rescaleFactor, long[] result, int offset)
+    {
+        if (rescaleFactor >= POWERS_OF_TEN.length) {
+            throwOverflowException();
+        }
+
+        boolean negative = high < 0;
+
+        if (negative) {
+            long tmpHigh = negateHighExact(high, low);
```

Comment:
No, that’s intentional. It’s a pattern I’ve been using to avoid creating intermediate arrays wherever possible. Sometimes the VM can elide those allocations，but often times it can’t, and it’s hard to predict.